### DISCUSSION TITLE
Go for beginners: Getting started

### DISCUSSION THREAD

**Discussion by jyyi1** [OutlineFoundation/outline-sdk#67] [Tutorials]
## Introduction

Outline SDK is a powerful Go[^1] library that can be used to build cross-platform VPN applications. However, getting started with the cross platform compilation can be a bit tricky, especially if you are new to Go.

This guide walks you through the process of setting up your Go environment and using the Outline SDK. It also covers how to create Go libraries that can be used on a variety of platforms, including Android, iOS/macOS, Windows and Linux.

If you have any questions about the environment setup, please feel free to post them here or in a new discussion thread.

## Set up Go[^1]

This section describes how to install Go. After your environment is set up, you will be able to create an executable that splits TCP streams into two streams: The first stream contains the first 3 bytes, and the second stream contains the rest of the data. You will be using the Outline SDK to create this executable.

### Install Go

You can follow the official installation guide at https://go.dev/doc/install [^2]. Alternatively, if you are using a package manager, you can also use it to install Go:

<details>
  <summary>macOS Homebrew</summary>

```sh
brew install go
```
[^3]
</details>
<details>
  <summary>Windows WinGet</summary>

```powershell
winget install -e --id GoLang.Go
```
[^4]
</details>
<details>
  <summary>Ubuntu apt</summary>

```sh
sudo add-apt-repository ppa:longsleep/golang-backports
sudo apt update
sudo apt install golang-go
```
[^5]
</details>

After Go is installed, you can verify that it is installed correctly by running the following command (make sure the version is ≥ `1.20`):

```sh
go version
```

If you want to learn some basic Go syntax, "A Tour of Go"[^6] is a great online tool.

### Create `splitfetch` Application

This section describes how to create a cross-platform `split-fetcher` application that will fetch a page using HTTP/HTTPS protocol. Instead of sending the entire request in a single network packet, it will split it into two packets. This helps the app to circumvent deep-packet inspection [^7]. This application will run on desktop platforms. For mobile usage, see the following sections.

To set up the `splitfetch` project, follow these steps:

<details>
<summary>Details</summary>

1. Create a folder to contain your code; initialize a Go project and add the Outline SDK dependency (this will add `go.mod` and `go.sum` file to this folder):
   ```sh
   mkdir splitfetch
   cd splitfetch
   go mod init local/example/splitfetch
   go get github.com/Jigsaw-Code/outline-sdk@latest
   ```
   > Note: If you need to push the code to a version control system such as a GitHub repository, we recommend you to use a fully qualified package name "github.com/your-org/your-repo" instead of the one used here.
2. Create two files: `fetcher.go` and `main/main.go`. `fetcher.go` will include a helper function that fetches a page using Outline SDK's [`split.NewStreamDialer`](https://github.com/Jigsaw-Code/outline-sdk/blob/main/transport/split/stream_dialer.go). This function will be used by `main.go`, which is the entrypoint of the application:
   ```sh
   touch fetcher.go
   mkdir main
   touch main/main.go
   ```
</details>

Then, use your favorite code editor[^8] to paste the code below into `fetcher.go` and `main/main.go`:

<details>
<summary>fetcher.go</summary>

```go
package splitfetch

import (
	"context"
	"fmt"
	"io"
	"net"
	"net/http"
	"strings"

	"github.com/Jigsaw-Code/outline-sdk/transport"
	"github.com/Jigsaw-Code/outline-sdk/transport/split"
)

const splitPacketSize = 3

func FetchPage(url string) (string, error) {
	dialer, err := split.NewStreamDialer(&transport.TCPDialer{}, split.NewFixedSplitIterator(splitPacketSize))
	if err != nil {
		return "", err
	}

	httpClient := &http.Client{
		Transport: &http.Transport{
			DialContext: func(ctx context.Context, network, addr string) (net.Conn, error) {
				if !strings.HasPrefix(network, "tcp") {
					return nil, fmt.Errorf("protocol not supported: %v", network)
				}
				return dialer.DialStream(ctx, addr)
			},
		},
	}

	resp, err := httpClient.Get(url)
	if err != nil {
		return "", err
	}
	defer resp.Body.Close()

	buf := &strings.Builder{}
	_, err = io.Copy(buf, resp.Body)
	if err != nil {
		return "", nil
	}
	return buf.String(), nil
}
```
</details>
<details>
<summary>main/main.go</summary>

```go
package main

import (
	"fmt"
	"local/example/splitfetch"
	"log"
	"os"
)

func main() {
	page, err := splitfetch.FetchPage(os.Args[1])
	if err != nil {
		log.Fatalf("%v", err)
	}
	fmt.Println(page)
}
```
</details>

After you have copied the code into the files, you can run `go mod tidy` to update the `go.mod` file.

### Run `splitfetch` Application

You can now run the `splitfetch` application by:

```sh
go run main/main.go https://ipinfo.io
```

This will compile and run the application. If you want to create a distributable binary, you can use the `go build` command:

<details>
<summary>Windows</summary>

```powershell
go build -o splitfetch.exe main/main.go
```
</details>
<details>
<summary>macOS or Linux</summary>

```sh
go build -o splitfetch main/main.go
```
</details>

## Create a Go library

In some cases, it is more desirable to use the `splitfetch.FetchPage` as a library rather than a standalone executable binary. For example, when developing an Android app.

This section describes the process of creating a reusable library for `splitfetch.FetchPage`. It demonstrates how to build this library for different platforms that can be readily integrated into your applications.

### Mobile Platforms

This tutorial uses Go Mobile[^9] to create the library for Android and iOS/macOS. First, you will need to install the `gomobile` command line tool [^10] using the following command:

```sh
go install golang.org/x/mobile/cmd/gomobile@latest
```

By default, it will be installed to the `"$(go env GOPATH)/bin"` folder (typically it expands to `"$HOME/go/bin"` on Linux/macOS, or `"%HOMEPATH%\go\bin"` on Windows). You can override this location by `export GOBIN="<customized-bin-folder>"` before running the `go install` command.

You must also add the bin folder mentioned above to the `PATH` environment variable, because `gomobile` will try to install and discover some additional command line tools:

<details>
<summary>Linux/macOS</summary>

```sh
export PATH="$PATH:$(go env GOPATH)/bin"
```
</details>
<details>
<summary>Windows (PowerShell)</summary>

```powershell
$Env:PATH+=";$(go env GOPATH)\bin"
```
</details>

After you have completed these steps, you can initialize the `gomobile` command line tools [^10]:

```sh
gomobile init
```

#### Build Android Library

To build an Android library (AAR, Android Archive) for the `FetchPage` function, you need to install the Android SDK (API Level 18+) and JDK (8+). We recommend installing the latest Android Studio [^11], which includes both of these requirements.

After you have installed Android Studio, check that JDK is installed correctly by running `javac --version` in the terminal. If you see any errors, install the JDK from a vendor such as Oracle[^12] or OpenJDK[^13]. It is also usually available in package managers such as brew[^14], WinGet[^15], and apt[^16].

You must also set the `ANDROID_HOME` environment variable to the location of your Android SDK. You can find this location in Android Studio [^17]:

```sh
export ANDROID_HOME="$HOME/Android/Sdk"    ## Replace it with your actual Android SDK path ##
```

You can now build the Android library using the following commands (for more `gomobile` options, see their documentation [^18]):

```sh
go get golang.org/x/mobile/bind
gomobile bind -target android -o splitfetch.aar .
```

> 💡 `golang.org/x/mobile/bind` is needed when running `gomobile`, but `go mod tidy` will remove it because it is not used by any Go source code. One way to fix this is to `import ( _ "golang.org/x/mobile" )` in a dedicated `tools.go` file [^24].

To use this library in your Android app, follow these steps:

1. Create a new project in Android Studio.
2. Import the `splitfetch.aar` file into your project [^19].
3. In the Manifest file [^20], declare the internet permission [^21].
4. Use the following Java code to call the `FetchPage` function:
    ```java
    import splitfetch.Splitfetch;
    // ... unrelated code omitted ...
    try {
        String resp = Splitfetch.fetchPage("https://ipinfo.io");
        // resp contains the response content string
    } catch (Exception ex) {
        // handle errors
    }
    ```

#### Build iOS/macOS Library

To create an XCFramework bundle [^22] for the `FetchPage` function, you need a macOS machine with Xcode[^23] installed. After Xcode[^23] is installed, you can build the XCFramework bundle using the following command (for more `gomobile` options, see their documentation [^18]):

```sh
go get golang.org/x/mobile/bind
gomobile bind -target=ios,macos -o splitfetch.xcframework .
```

> 💡 `golang.org/x/mobile/bind` is needed when running `gomobile`, but `go mod tidy` will remove it because it is not used by any Go source code. One way to fix this is to `import ( _ "golang.org/x/mobile" )` in a dedicated `tools.go` file [^24].

To use this multiplatform XCFramework library in your iOS or macOS app, you must first create a Swift Package [^25] for the binary [^26]. To do this:

1. Create a new "Swift Package" project in Xcode (e.g. "SplitFetchLib").
2. Modify the `Package.swift` file:
    ```swift
    // ... unchanged ...
    let package = Package(
        // ... unchanged ...
        products: [
            .library(
                name: "SplitFetchLib",
                targets: ["SplitFetchLib", "splitfetch"]),  // add "splitfetch" target
        ],
        // ... unchanged ...
        targets: [
            // ... unchanged ...
            // add the xcframework binary (must use relative path)
            .binaryTarget(
                name: "splitfetch",
                path: "<path-relative-to-this-swift-file>/splitfetch.xcframework"),
        ]
    )
    ```
3. `Product > Build` this project to make sure everything is good.

Next, close the package project. The steps below create an app to use the package:

1. Create a new "App" project in Xcode (e.g. "FetchingApp").
2. Add the "SplitFetchLib" dependency to the project: `File > Add Packages... > Add Local... > choose "SplitFetchLib" folder > Add Package` [^27].
3. Make sure "SplitFetchLib" is included in the build process: `select project in the left nav pane > under "TARGETS", select project > General Pane > Frameworks, Libraries, and Embedded Content > + > select the leaf node of "SplitFetchLib" > Add`.
4. Enable internet capability by checking the `Outgoing Connections (Client)` checkbox on the `Signing & Capabilities` page of the project. [^28][^29]
5. Use the following Swift code to call the `FetchPage` function:
    ```swift
    import Splitfetch
    // ... unrelated code omitted ...
    let err: NSErrorPointer = nil
    let response = SplitfetchFetchPage("https://ipinfo.io", err)
    if err != nil {
        // handle errors
    }
    ```
6. `Product > Build` to make sure everything is good. If you get any errors, double-check step 2 and step 3 above. Sometimes `File > Packages > Reset Package Caches` and `Product > Clean Build Folder` might also help.

> 💡 If you don't need a multiplatform app, you can use the library even more easily (no need to create a Swift package wrapper): just drag a corresponding `Splitfetch.framework` folder in the `splitfetch.xcframework` to your project, and call the `FetchPage` function as specified in Step 5 above. Here is an example [^30].

### For Desktop Platforms

The best way to consume this package on desktop platforms such as Linux and Windows is to use Go directly (you can even write GUIs in Go [^42]). However, if this is not possible (for example, if you have an existing application written in another language, or if you are creating an embedded application), you can still create a C library using  `cgo`[^31][^32]. The C library is a glue that can be used to connect different programming languages together [^33][^34][^35][^36]. You need to install a native C/C++ compiler on your platform before using `cgo`. For example, on Linux you need to install `gcc`/`g++` [^37][^38], on Windows you need to install MinGW-w64 [^39][^44] (MSVC is not currently supported [^43]), and on macOS you need to install Xcode [^23].

> 💡 Using a native C/C++ compiler means you can only compile your program to your computer's platform. It is possible to cross-compile to different platforms, but it is more complicated and is not covered in this document. If you want to cross-compile, you need to install a cross-compile C/C++ compiler (such as llvm/clang[^40]) and have the target platform's sysroot [^41] ready.

To compile this package to a C library, you need to add another main package that defines C-compatible functions (you can be more creative than us when defining the parameters and return type of the functions, see more in `cgo` documentation [^32]):

```sh
mkdir clib
touch clib/main.go
```

<details>
<summary>clib/main.go</summary>

```go
package main

// #include <stdlib.h>
import "C"
import (
	"local/example/splitfetch"
	"unsafe"
)

//export fetch_page
func fetch_page(url *C.char) *C.char {
	resp, err := splitfetch.FetchPage(C.GoString(url))
	if err != nil {
		resp = err.Error()
	}
	return C.CString(resp)
}

//export free_string
func free_string(s *C.char) {
	C.free(unsafe.Pointer(s))
}

func main() { /* dummy to make `go build` happy */ }
```
</details>

After you have completed these steps, you will be able to build it into a C shared library:

<details>
<summary>Linux/macOS</summary>

```sh
go build -buildmode c-shared -o lib/splitfetch.so ./clib/main.go
```
</details>
<details>
<summary>Windows (PowerShell)</summary>

```powershell
# make sure gcc.exe is available through PATH environment variable
go build -buildmode c-shared -o lib/splitfetch.dll ./clib/main.go
```
</details>

After running the previous command, you will have two files in the `lib` folder: a shared library binary (for example, `splitfetch.so`/`splitfetch.dll`) and a header file (`splitfetch.h`).

#### Use the shared library (C/C++)

This section creates a native C++ application to demonstrate how to use this shared library. Create a new folder and copy both files (i.e. `splitfetch.so`/`splitfetch.dll` and `splitfetch.h`) into it. Then create a C file called `main.cpp` in this folder.

<details>
<summary>main.cpp</summary>

```cpp
#include <iostream>

#include "splitfetch.h"

int main(void)
{
    char *resp = fetch_page(const_cast<char*>("https://ipinfo.io"));
    std::cout << "Result:\n" << resp << std::endl;
    free_string(resp);
    return 0;
}
```
</details>

To compile and link the C++ code with the shared library, and run the final executable file (it's important to make sure that the shared library and the executable file are in the same folder):

<details>
<summary>Linux/macOS</summary>

```sh
g++ -o fetching main.cpp splitfetch.so
./fetching
```
</details>
<details>
<summary>Windows (PowerShell)</summary>

```sh
# make sure g++.exe is available through environment variable
g++ -o fetching main.cpp splitfetch.dll
./fetching.exe
```
</details>

#### Use the shared library (C#)

This section demonstrates how to use this shared library in C#. First, you will need to have the .NET SDK [^45] installed on your system. Then you can create, build and run your C# application:

```powershell
dotnet new console -n fetcher
cd fetcher
# copy the file content below into Program.cs
dotnet build -o ./bin
# copy "splitfetch.so"/"splitfetch.dll" to "./bin" folder
# (you do not need to copy the header file)
./bin/fetcher.exe
```

<details>
<summary>Program.cs</summary>

```csharp
using System.Runtime.InteropServices;

[DllImport("splitfetch")]
static extern string fetch_page(string url);

[DllImport("splitfetch")]
static extern void free_string(string s);

var result = fetch_page("https://ipinfo.io");
Console.WriteLine(result);
free_string(result);
```
</details>

#### Footnotes

[^1]: [The Go Programming Language](https://go.dev/)
[^2]: [Download and Install Go](https://go.dev/doc/install)
[^3]: [Homebrew: go](https://formulae.brew.sh/formula/go)
[^4]: [WinGet: Golang.Go.1.20](https://winget.run/pkg/GoLang/Go.1.20)
[^5]: [Install Go on Ubuntu](https://github.com/golang/go/wiki/Ubuntu)
[^6]: [A Tour of Go](https://go.dev/tour/welcome/1)
[^7]: [New Feature: Packet Splitting](https://github.com/Jigsaw-Code/outline-sdk/discussions/62)
[^8]: [Go Editors and IDEs](https://go.dev/doc/editors)
[^9]: [Go Mobile](https://github.com/golang/go/wiki/Mobile)
[^10]: [`gomobile` CLI](https://pkg.go.dev/golang.org/x/mobile/cmd/gomobile)
[^11]: [Install Android Studio](https://developer.android.com/studio/install)
[^12]: [Oracle JDK 21](https://www.oracle.com/java/technologies/downloads/#java21)
[^13]: [Open JDK 21](https://jdk.java.net/21/)
[^14]: [`brew install openjdk`](https://formulae.brew.sh/formula/openjdk)
[^15]: [WinGet Oracle JDK 17](https://winget.run/pkg/Oracle/JDK.17)
[^16]: [`apt install default-jdk`](https://www.digitalocean.com/community/tutorials/how-to-install-java-with-apt-on-ubuntu-20-04)
[^17]: [Find Android SDK Location](https://developer.android.com/tools#:~:text=In%20Android%20Studio%2C%20click%20File,shown%20under%20Android%20SDK%20location.)
[^18]: [`gomobile` Build a library for Android and iOS](https://pkg.go.dev/golang.org/x/mobile/cmd/gomobile#hdr-Build_a_library_for_Android_and_iOS)
[^19]: [Android Studio: add AAR or JAR as a dependency](https://developer.android.com/studio/projects/android-library#psd-add-aar-jar-dependency)
[^20]: [Android Studio: declaring app permissions](https://developer.android.com/training/permissions/declaring)
[^21]: [Network permissions for Android app](https://developer.android.com/training/basics/network-ops/connecting)
[^22]: [Apple: multiplatform binary framework bundle](https://developer.apple.com/documentation/xcode/creating-a-multi-platform-binary-framework-bundle)
[^23]: [Xcode on the Mac App Store](https://apps.apple.com/us/app/xcode/id497799835?mt=12)
[^24]: [Go Modules: how can I track tool dependencies for a module](https://github.com/golang/go/wiki/Modules#how-can-i-track-tool-dependencies-for-a-module)
[^25]: [Creating a standalone Swift package with Xcode](https://developer.apple.com/documentation/xcode/creating-a-standalone-swift-package-with-xcode)
[^26]: [Distributing binary frameworks as Swift packages](https://developer.apple.com/documentation/xcode/distributing-binary-frameworks-as-swift-packages)
[^27]: [Adding package dependencies to your app](https://developer.apple.com/documentation/xcode/adding-package-dependencies-to-your-app)
[^28]: [Adding capabilities to your app](https://developer.apple.com/documentation/xcode/adding-capabilities-to-your-app/)
[^29]: [Configuring the macOS App Sandbox](https://developer.apple.com/documentation/xcode/configuring-the-macos-app-sandbox)
[^30]: [Calling Go code from Swift on iOS and vice versa with Gomobile](https://medium.com/@matryer/tutorial-calling-go-code-from-swift-on-ios-and-vice-versa-with-gomobile-7925620c17a4)
[^31]: [C? Go? Cgo!](https://go.dev/blog/cgo)
[^32]: [cgo command](https://pkg.go.dev/cmd/cgo)
[^33]: [NodeJS: C++ addons](https://nodejs.org/api/addons.html)
[^34]: [C#: Platform Invoke](https://learn.microsoft.com/en-us/dotnet/standard/native-interop/pinvoke)
[^35]: [Extending Python with C or C++](https://docs.python.org/3/extending/extending.html)
[^36]: [Java: Guide to JNI](https://www.baeldung.com/jni)
[^37]: [Installing GCC](https://gcc.gnu.org/install/)
[^38]: [How to Install GCC Compiler on Ubuntu](https://phoenixnap.com/kb/install-gcc-ubuntu)
[^39]: [MinGW-w64 Downloads](https://www.mingw-w64.org/downloads/)
[^40]: [LLVM/Clang](https://clang.llvm.org/)
[^41]: [Cross compiling made easy, using Clang and LLVM](https://mcilloni.ovh/2021/02/09/cxx-cross-clang/)
[^42]: [A list of Go GUI projects](https://github.com/go-graphics/go-gui-projects#a-list-of-go-guigraphicsimage-related-projects)
[^43]: [Go cmd/link: support msvc object files](https://github.com/golang/go/issues/20982)
[^44]: [MSYS2](https://www.msys2.org/)
[^45]: [Install .NET on Windows, Linux, and macOS](https://learn.microsoft.com/en-us/dotnet/core/install/)

**Comment by SiteSplat** [comment]
Waiting for the: "For Desktop Platforms" 🙏

**Reply by maddyhof** [reply]
Coming soon!! Sorry for the delay
